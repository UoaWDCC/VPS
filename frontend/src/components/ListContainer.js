import React, { useState } from "react";
import ImageList from "@material-ui/core/ImageList";
import ImageListItem from "@material-ui/core/ImageListItem";
import { Box } from "@material-ui/core";
import Thumbnail from "./Thumbnail";
import styles from "../styling/ListContainer.module.scss";
import DashedCard from "./DashedCard";
import useStyles from "./component.styles";

/**
 * Component used to display cards in a list format for scenario and scene selection.
 *
 * @component
 * @example
 * const data = [ ... ]
 * const wide = true
 * const sceneSelectionPage = false
 * const scenarioId = "1ef4cD1wsd676dS"
 * function onItemSelected() {
 *   console.log("Selected.")
 * }
 * function addCard() {
 *   console.log("Card Added.")
 * }
 * function onItemBlur() {
 *   console.log("Item Blurred."")
 * }
 * return (
 *   <ListContainer
 *     data={data}
 *     wide={wide}
 *     sceneSelectionPage={sceneSelectionPage}
 *     scenarioId={scenarioId}
 *     onItemSelected={onItemSelected}
 *     addCard={addCard}
 *     onItemBlur={onItemBlur}
 *   />
 * )
 */
export default function ListContainer({
  data,
  onItemSelected,
  wide,
  addCard,
  onItemBlur,
  sceneSelectionPage,
  scenarioId,
  invalidName,
}) {
  const classes = useStyles();
  const [selected, setSelected] = useState();
  const columns = wide ? 5 : 4;

  /** Function which executes when an image in the image list is clicked. */
  const onItemClick = (item) => {
    setSelected(item._id);
    onItemSelected(item);
  };

  return (
    <>
      <div
        className={
          wide ? styles.scenarioListContainerWide : styles.scenarioListContainer
        }
      >
        <ImageList rowHeight={210} cols={columns} gap={30}>
          {addCard ? (
            <ImageListItem
              className={classes.listContainerItem}
              key={-1}
              cols={1}
              height={200}
            >
              <DashedCard onClick={addCard} />
            </ImageListItem>
          ) : null}
          {data && data.length > 0
            ? data.map((item) => (
                <ImageListItem
                  className={classes.listContainerItem}
                  key={item._id}
                  cols={1}
                  height={200}
                  onClick={() => onItemClick(item)}
                >
                  <div
                    className={
                      wide ? styles.imageListItemWide : styles.imageListItem
                    }
                  >
                    <Box
                      height={160}
                      border={5}
                      borderRadius={10}
                      borderColor={
                        item._id === selected ? "#008a7b" : "#747474"
                      }
                      overflow="hidden"
                      textAlign="center"
                      sx={{
                        background: "#f1f1f1",
                        "&:hover": {
                          background: "#cccccc",
                        },
                      }}
                    >
                      {sceneSelectionPage ? (
                        <Thumbnail
                          url={`${process.env.PUBLIC_URL}/play/${scenarioId}/${item._id}`}
                        />
                      ) : (
                        <Thumbnail
                          url={`${process.env.PUBLIC_URL}/play/${item._id}`}
                        />
                      )}
                    </Box>
                    <input
                      className={styles.text}
                      defaultValue={item.name}
                      onBlur={onItemBlur}
                      key={item._id}
                    />
                  </div>
                  {invalidName && (
                    <p1 className="nullNameWarning">invalid null name</p1>
                  )}
                </ImageListItem>
              ))
            : null}
        </ImageList>
      </div>
    </>
  );
}
