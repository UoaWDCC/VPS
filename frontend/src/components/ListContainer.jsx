import { useState } from "react";

import { Box } from "@material-ui/core";
import ImageList from "@material-ui/core/ImageList";
import ImageListItem from "@material-ui/core/ImageListItem";

import Thumbnail from "containers/pages/AuthoringTool/Components/Thumbnail";
import DashedCard from "./DashedCard";

import styles from "../styling/ListContainer.module.scss";
import useStyles from "./component.styles";

/**
 * Component used to display cards in a list format for scenario and scene selection.
 *
 * @component
 * @example
 * const data = [ ... ]
 * const wide = true
 * const sceneSelectionPage = false
 * const scenarioId = "1ef4cD1wsd676dS"
 * function onItemSelected() {
 *   console.log("Selected.")
 * }
 * function onItemDoubleClick() {
 *   console.log("Double clicked.")
 * }
 * function addCard() {
 *   console.log("Card Added.")
 * }
 * function onItemBlur() {
 *   console.log("Item Blurred.")
 * }
 * return (
 *   <ListContainer
 *     data={data}
 *     wide={wide}
 *     sceneSelectionPage={sceneSelectionPage}
 *     scenarioId={scenarioId}
 *     onItemSelected={onItemSelected}
 *     onItemDoubleClick={onDoubleClick}
 *     addCard={addCard}
 *     onItemBlur={onItemBlur}
 *   />
 * )
 */
export default function ListContainer({
  data, // could be scenarios or scenes data
  assignedScenarios,
  onItemSelected,
  onItemDoubleClick,
  wide,
  addCard,
  onItemBlur,
  sceneSelectionPage,
  scenarioId,
  invalidNameId,
}) {
  const classes = useStyles();
  const [selected, setSelected] = useState();
  const columns = wide ? 5 : 4;

  /** Function which executes when an image in the image list is clicked. */
  const onItemClick = (event, item) => {
    if (event.detail === 2) {
      onItemDoubleClick(item);
    } else {
      setSelected(item._id);
      onItemSelected(item);
    }
  };

  /** Function which executes when an image in the image list is right-clicked. Select item. */
  const onItemRightClick = (item) => {
    setSelected(item._id);
    onItemSelected(item);
  };

  return (
    <>
      <div
        className={
          wide ? styles.scenarioListContainerWide : styles.scenarioListContainer
        }
      >
        {!sceneSelectionPage && (
          <h1 className="text-3xl font-bold my-3">Created scenarios</h1>
        )}

        <ImageList rowHeight={210} cols={columns} gap={30}>
          {addCard ? (
            <ImageListItem
              className={classes.listContainerItem}
              key={-1}
              cols={1}
              height={200}
            >
              <DashedCard onClick={addCard} />
            </ImageListItem>
          ) : null}
          {data && data.length > 0
            ? data.map((item) => (
                <ImageListItem
                  className={classes.listContainerItem}
                  key={item._id}
                  cols={1}
                  height={200}
                  onClick={(event) => onItemClick(event, item)}
                  onContextMenu={() => onItemRightClick(item)}
                >
                  <div
                    className={
                      wide ? styles.imageListItemWide : styles.imageListItem
                    }
                  >
                    <Box
                      height={160}
                      border={5}
                      borderRadius={10}
                      borderColor={
                        item._id === selected ? "#035084" : "#747474"
                      }
                      overflow="hidden"
                      textAlign="center"
                      display="flex"
                      sx={{
                        background: "#f1f1f1",
                        "&:hover": {
                          background: "#cccccc",
                        },
                      }}
                    >
                      {sceneSelectionPage ? (
                        <Thumbnail components={item.components} />
                      ) : (
                        <Thumbnail
                          components={item.thumbnail?.components || []}
                        />
                      )}
                    </Box>
                    <input
                      className={styles.text}
                      defaultValue={item.name}
                      onBlur={onItemBlur}
                      key={item._id}
                    />
                  </div>
                  {invalidNameId === item._id && (
                    <p1 className="nullNameWarning">invalid null name</p1>
                  )}
                </ImageListItem>
              ))
            : null}
        </ImageList>

        {assignedScenarios ? (
          <>
            {!sceneSelectionPage && (
              <h1 className="text-3xl font-bold my-3">Assigned scenarios</h1>
            )}

            <ImageList rowHeight={210} cols={columns} gap={30}>
              {addCard ? (
                <ImageListItem
                  className={classes.listContainerItem}
                  key={-1}
                  cols={1}
                  height={200}
                >
                  <DashedCard onClick={addCard} />
                </ImageListItem>
              ) : null}
              {assignedScenarios && assignedScenarios.length > 0
                ? assignedScenarios.map(({ _id, name }) => (
                    <ImageListItem
                      className={classes.listContainerItem}
                      key={_id}
                      cols={1}
                      height={200}
                      onClick={(event) => onItemClick(event, { _id })}
                      onContextMenu={() => onItemRightClick({ _id })}
                    >
                      <div
                        className={
                          wide ? styles.imageListItemWide : styles.imageListItem
                        }
                      >
                        <Box
                          height={160}
                          border={5}
                          borderRadius={10}
                          borderColor={_id === selected ? "#035084" : "#747474"}
                          overflow="hidden"
                          textAlign="center"
                          sx={{
                            background: "#f1f1f1",
                            "&:hover": {
                              background: "#cccccc",
                            },
                          }}
                        >
                          {sceneSelectionPage ? (
                            <Thumbnail url={`/play/${scenarioId}/${_id}`} />
                          ) : (
                            <Thumbnail url={`/play/${_id}`} />
                          )}
                        </Box>
                        <input
                          className={styles.text}
                          defaultValue={name}
                          onBlur={onItemBlur}
                          key={_id}
                        />
                      </div>
                      {invalidNameId === _id && (
                        <p1 className="nullNameWarning">invalid null name</p1>
                      )}
                    </ImageListItem>
                  ))
                : null}
            </ImageList>
          </>
        ) : null}
      </div>
    </>
  );
}
