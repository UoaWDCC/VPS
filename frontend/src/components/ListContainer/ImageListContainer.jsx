import { Box } from "@material-ui/core";
import ImageList from "@material-ui/core/ImageList";
import ImageListItem from "@material-ui/core/ImageListItem";

/**
 * Component used to display images in a list format.
 */
export default function ImageListContainer({ data, onItemSelected, selectedId }) {
  const rowHeight = 150;

  const onItemClick = (item) => {
    onItemSelected(item);
  };

  return (
    <ImageList rowHeight={rowHeight} cols={2} gap={10}>
      {data?.map((item) => (
        <ImageListItem
          key={item.id || item._id} // fallback if some have _id instead
          cols={1}
          onClick={() => onItemClick(item)}
        >
          <Box
            height={rowHeight}
            minWidth={rowHeight}
            border={item.id === selectedId ? 5 : 0}
            borderColor={item.id === selectedId ? "#00b0e6" : "#747474"}
            overflow="hidden"
            sx={{
              "&:hover": {
                opacity: "0.5",
                cursor: "pointer",
              },
              backgroundImage: `url(${item.url})`,
              backgroundSize: "cover",
              backgroundPosition: "center",
              boxSizing: "border-box",
            }}
          />
        </ImageListItem>
      ))}
    </ImageList>
  );
}
