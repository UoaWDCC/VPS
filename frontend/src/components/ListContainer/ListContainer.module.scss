.scenarioListContainer {
  width: 85vw;
  height: 100vh;
  overflow-y: scroll;
  overflow-x: hidden;
  padding-right: 30px;
  padding-left: 30px;
}

.scenarioListContainerWide {
  flex-grow: 1;
  width: 100vw;
  height: 90vh;
  overflow-y: scroll;
  overflow-x: hidden;
}

.imageListItem {
  margin-top: 20px;
  margin-left: 3px;
  margin-right: 3px;
  cursor: pointer;
}

.imageListItemWide {
  padding-top: 20px;
  padding-left: 20px;
  padding-right: 20px;
  cursor: pointer;
}

.crossHorizontalLine {
  position: absolute;
  top: calc(50% - 5px);
  left: 0;
  right: 0;
  height: 10px;
  background: #adadad;
  width: 80px;
  margin: 0 auto;
  border-radius: 4px;
}

.crossVerticalLine {
  position: absolute;
  left: calc(50% - 5px);
  top: 0;
  bottom: 0;
  width: 10px;
  background: #adadad;
  height: 80px;
  margin: auto 0;
  border-radius: 4px;
}

.text {
  display: inline-block;
  font-family: "Lucida Sans", "Lucida Sans Regular", "Lucida Grande",
    "Lucida Sans Unicode", Geneva, Verdana, sans-serif;
  text-align: center;
  width: 60%;
  border: 2px solid transparent;
  box-sizing: border-box;
  margin: auto;
  margin-top: 5px;

  &:focus {
    outline: none;
    border: 2px solid #0080a7;
    border-radius: 5px;
  }
}
