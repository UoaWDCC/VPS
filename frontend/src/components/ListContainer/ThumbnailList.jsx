import { useState } from "react";

import { Box } from "@material-ui/core";
import ImageList from "@material-ui/core/ImageList";
import ImageListItem from "@material-ui/core/ImageListItem";

import Thumbnail from "features/authoring/components/Thumbnail";
import DashedCard from "../DashedCard";

import styles from "./ThumbnailList.module.scss";

/**
 * Component used to display cards in a list format for scenario and scene selection.
 *
 * @component
 * @example
 * const data = [ ... ]
 * function onItemSelected() {
 *   console.log("Selected.")
 * }
 * function onItemDoubleClick() {
 *   console.log("Double clicked.")
 * }
 * function addCard() {
 *   console.log("Card Added.")
 * }
 * function onItemBlur() {
 *   console.log("Item Blurred.")
 * }
 * return (
 *   <ListContainer
 *     data={data}
 *     onItemSelected={onItemSelected}
 *     onItemDoubleClick={onDoubleClick}
 *     addCard={addCard}
 *     onItemBlur={onItemBlur}
 *   />
 * )
 */
export default function ThumbnailList({
  data, // could be scenarios or scenes data, but expects components.
  invalidNameId,
  highlightOnSelect = true, // Whether or not to highlight the card border on select.
  addCard,
  onItemBlur,
  onItemSelected = () => {},
  onItemDoubleClick = () => {},
}) {
  const [selected, setSelected] = useState();

  /** Function which executes when an image in the image list is clicked. */
  const onItemClick = (event, item) => {
    if (event.detail === 2) {
      onItemDoubleClick(item);
    } else {
      if (highlightOnSelect) {
        setSelected(item._id);
      }
      onItemSelected(item);
    }
  };

  /** Function which executes when an image in the image list is right-clicked. Select item. */
  const onItemRightClick = (item) => {
    setSelected(item._id);
    onItemSelected(item);
  };

  return (
    <>
      <div className={styles.scenarioListContainer}>
        <ImageList rowHeight={210} gap={30}>
          {addCard ? (
            <ImageListItem
              className="min-w-72 max-w-80 min-h-48"
              key={-1}
              cols={1}
            >
              <DashedCard onClick={addCard} />
            </ImageListItem>
          ) : null}
          {data && data.length > 0
            ? data.map((item) => (
                <ImageListItem
                  className="min-w-72 max-w-80 min-h-48"
                  key={item._id}
                  cols={1}
                  onClick={(event) => onItemClick(event, item)}
                  onContextMenu={() => onItemRightClick(item)}
                >
                  <div className="flex flex-col gap-2">
                    <Box
                      className="cursor-pointer"
                      height={160}
                      border={item._id === selected ? 4 : 2}
                      borderRadius={10}
                      borderColor={
                        item._id === selected ? "#035084" : "#94a3b8 "
                      }
                      overflow="hidden"
                      textAlign="center"
                      display="flex"
                      justifyContent="center"
                      sx={{
                        background: "#e2e8f0",
                        "&:hover": {
                          background: "#035084",
                        },
                      }}
                    >
                      <Thumbnail components={item?.components || []} />
                    </Box>
                    <div className="w-full flex justify-center ">
                      {onItemBlur ? (
                        <input
                          className="w-fit font-mona bg-white border border-slate-300 rounded-lg text-center px-2 py-[0.125rem] max-w-full overflow-ellipsis"
                          defaultValue={item.name}
                          onBlur={onItemBlur}
                          key={item._id}
                        />
                      ) : (
                        <p className="w-fit font-mona bg-slate-200 rounded-full text-center px-4 py-[0.125rem] max-w-full overflow-ellipsis">
                          {item.name}
                        </p>
                      )}
                    </div>
                  </div>
                  {invalidNameId === item._id && (
                    <p1 className="nullNameWarning">invalid null name</p1>
                  )}
                </ImageListItem>
              ))
            : null}
        </ImageList>
      </div>
    </>
  );
}
