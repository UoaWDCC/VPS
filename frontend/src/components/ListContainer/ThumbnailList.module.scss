.listContainer {
  width: 100%;
  overflow-y: scroll;
  overflow-x: hidden;
  padding-right: 30px;
  padding-left: 30px;
}

.imageListItem {
  margin-top: 20px;
  margin-left: 3px;
  margin-right: 3px;
  cursor: pointer;
}

.crossHorizontalLine {
  position: absolute;
  top: calc(50% - 5px);
  left: 0;
  right: 0;
  height: 10px;
  background: #adadad;
  width: 80px;
  margin: 0 auto;
  border-radius: 4px;
}

.crossVerticalLine {
  position: absolute;
  left: calc(50% - 5px);
  top: 0;
  bottom: 0;
  width: 10px;
  background: #adadad;
  height: 80px;
  margin: auto 0;
  border-radius: 4px;
}

.text {
  display: inline-block;
  font-family: "Lucida Sans", "Lucida Sans Regular", "Lucida Grande",
    "Lucida Sans Unicode", Geneva, Verdana, sans-serif;
  text-align: center;
  width: 60%;
  border: 2px solid transparent;
  box-sizing: border-box;
  margin: auto;
  margin-top: 5px;

  &:focus {
    outline: none;
    border: 2px solid #0080a7;
    border-radius: 5px;
  }
}
