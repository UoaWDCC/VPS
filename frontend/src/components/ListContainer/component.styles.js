import { makeStyles } from "@material-ui/core";

/**
 * This file contains all the styles used to override material-ui components which are being used within a component.
 */

const useStyles = makeStyles({});

export default useStyles;
