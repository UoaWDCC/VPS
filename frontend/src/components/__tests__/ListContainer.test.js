import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import ListContainer from "../ListContainer";

test("ListContainer component snapshot test", () => {
  const testData = [
    {
      _id: 1,
      name: "Scenario 1",
      img: "",
    },
  ];
  const component = renderer.create(
    <BrowserRouter>
      <ListContainer data={testData} />
    </BrowserRouter>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
