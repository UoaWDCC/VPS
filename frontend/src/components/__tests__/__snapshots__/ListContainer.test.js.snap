// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListContainer component snapshot test 1`] = `
<div
  className="scenarioListContainer"
>
  <ul
    className="MuiImageList-root"
    style={
      Object {
        "margin": -15,
      }
    }
  >
    <li
      className="MuiImageListItem-root"
      height={200}
      onClick={[Function]}
      style={
        Object {
          "height": 240,
          "padding": 15,
          "width": "25%",
        }
      }
    >
      <div
        className="MuiImageListItem-item"
      >
        <div
          className="imageListItem"
        >
          <div
            className="MuiBox-root MuiBox-root-1"
          />
          <p
            className="text"
          >
            Scenario 1
          </p>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
